conductor/lib/uploader.py

```diff
@@ -295,6 +374,83 @@ def do_upload(self, upload_url, filename, md5):
+        once all parts are uploaded. On successful part upload, response headers will contain an ETag.
+        This value must be tracked along with the part number in order to complete and hydrate the file.
+        """
+        uploads = []
+        complete_payload = {
+            "uploadID": upload["uploadID"],
+            "hash": md5,
+            "completedParts": [],
+            "project": self.project
+        }
+
+        # iterate over parts and upload
+        for part in upload["parts"]:
+            resp_headers = self._do_multipart_upload(
+                upload_url=part["url"],
+                filename=filename,
+                part_number=part["partNumber"],
+                part_size=upload["partSize"]
+            )
+
+            if resp_headers:
+                uploads.append(upload["uploadID"])
+                completed_part = {
+                    "partNumber": part["partNumber"],
+                    "etag": resp_headers['ETag'].strip('"')
+                }
+                complete_payload["completedParts"].append(completed_part)
+
+        # Complete multipart upload in order to hydrate file for availability
+        uri_path = '/api/v2/files/multipart/complete'
+        headers = {'Content-Type': 'application/json'}
+        self.api_client.make_request(uri_path=uri_path,
+                                     verb='POST',
+                                     headers=headers,
+                                     data=json.dumps(complete_payload),
+                                     raise_on_error=True,
+                                     use_api_key=True)
+
+        return uploads
+
+    def _do_multipart_upload(self, upload_url, filename, part_number, part_size):
+        with open(filename, 'rb') as fh:
+            # seek to the correct part position
+            start = (part_number - 1) * part_size
+            fh.seek(start)
+
+            # read up to part size determined by file-api
+            data = fh.read(part_size)
+            content_length = len(data)
+
+            # upload part
+            response = self.api_client.make_prepared_request(
+                verb="PUT",
+                url=upload_url,
+                headers={
+                    'Content-Type': 'application/octet-stream',
+                    'Content-Length': str(content_length),
+                },
+                params=None,
+                data=data,
+                tries=1,
+                remove_headers_list=["Transfer-Encoding"]  # s3 will return a 501 if the Transfer-Encoding header exists
+            )
+
+            # report upload progress
+            self.metric_store.increment('bytes_uploaded', content_length, filename)
+
+            # close response object to add back to pool
+            # https://requests.readthedocs.io/en/master/user/advanced/#body-content-workflow
+            response.close()
+
+            return response.headers
```

Comment:
Don't bother changing it, but IMO having this function return the full response object (rather than just the headers) feels a little more appropriate for a function whose responsibility is to perform a request.  Obviously you only need the headers, so it really doesn't matter.